### PR TITLE
merge string prs

### PR DESCRIPTION
**Pull request by stuartarchibald**
This is a merge-in/cherry-pick (as convenient) PR of the following:

#4823
#4824
#4825
#4834
#4842
#4843
#4844
#4845
#4849
#4865
#4867
#4868

With some refactoring as needed. There's more scope for additional refactoring. This PR is to make review faster.

-- 
Closes #4823
Closes #4824
Closes #4825
Closes #4834
Closes #4842
Closes #4843
Closes #4844
Closes #4845
Closes #4849 
Closes #4865
Closes #4867
Closes #4868

### CODE REVIEW

File: numba/unicode.py [hunk 1]
```diff
@@ -1582,110 +1955,118 @@ def impl(a):
...
-        return cased
-    return impl
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L12096-L12124    # noqa: E501
+overload_method(types.UnicodeType, 'isnumeric')(gen_isX(_PyUnicode_IsNumeric))
 
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L12056-L12085    # noqa: E501
+overload_method(types.UnicodeType, 'isdigit')(gen_isX(_PyUnicode_IsDigit))
+
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L12017-L12045    # noqa: E501
+overload_method(
```
Comment (sklam) [inline numba/unicode.py]: The formatting here is awkward. perhaps something like the below is better:

```python
overload_method(types.UnicodeType, 'isdecimal')(
    gen_isX(_PyUnicode_IsDecimalDigit),
)
```

File: numba/unicode.py [hunk 2]
```diff
@@ -1582,110 +1955,118 @@ def impl(a):
...
+    types.UnicodeType,
+    'isdecimal')(
+        gen_isX(_PyUnicode_IsDecimalDigit))
 
 # https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L12188-L12213    # noqa: E501
-@overload_method(types.UnicodeType, 'isprintable')
-def unicode_isprintable(data):
-    """Implements UnicodeType.isprintable()"""
+overload_method(
```
Comment (sklam) [inline numba/unicode.py]: Same here. It doesn't need to take that many lines.

File: numba/unicode.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/unicode.py]: I have to confess to just running an auto-formatter after doing all the merging/resolving conflicts. I've fixed this in d2584f6.

**Comment by stuartarchibald** [comment]
Think this is just waiting on an internal CI run for merge approval?

**Comment by sklam** [comment]
i was just waiting for public CI to finish.

**Pull request merged**